python_modules/dagster/dagster/_daemon/auto_run_reexecution/auto_run_reexecution.py

```diff
@@ -25,66 +19,10 @@ def filter_runs_to_should_retry(
     runs: Sequence[DagsterRun], instance: DagsterInstance, default_max_retries: int
 ) -> Iterator[Tuple[DagsterRun, int]]:
     """Return only runs that should retry along with their retry number (1st retry, 2nd, etc.)."""
-
-    def get_retry_number(run: DagsterRun) -> Optional[int]:
-        if run.status != DagsterRunStatus.FAILURE:
-            return None
-
-        raw_max_retries_tag = run.tags.get(MAX_RETRIES_TAG)
-        if raw_max_retries_tag is None:
-            max_retries = default_max_retries
-        else:
-            try:
-                max_retries = int(raw_max_retries_tag)
-            except ValueError:
-                instance.report_engine_event(
-                    f"Error parsing int from tag {MAX_RETRIES_TAG}, won't retry the run.", run
-                )
-                return None
-
-        if max_retries == 0:
-            return None
-
-        # TODO: group these to reduce db calls
-        run_group = instance.get_run_group(run.run_id)
-
-        if run_group:
-            _, run_group_iter = run_group
-            run_group_list = list(run_group_iter)
-
-            # Has the parent run already been retried the maximum number of times? (Group includes the parent)
-            if len(run_group_list) >= max_retries + 1:
-                return None
-
-            # Does this run already have a child run?
-            if any([run.run_id == run_.parent_run_id for run_ in run_group_list]):
-                return None
-            return len(run_group_list)
-        else:
-            return 1
-
-    default_retry_on_asset_or_op_failure: bool = instance.get_settings("run_retries").get(
-        "retry_on_asset_or_op_failure", True
-    )
-
     for run in runs:
-        retry_number = get_retry_number(run)
-        retry_on_asset_or_op_failure = get_boolean_tag_value(
-            run.tags.get(RETRY_ON_ASSET_OR_OP_FAILURE_TAG),
-            default_value=default_retry_on_asset_or_op_failure,
-        )
-        if retry_number is not None:
-            if (
-                run.tags.get(RUN_FAILURE_REASON_TAG) == RunFailureReason.STEP_FAILURE.value
-                and not retry_on_asset_or_op_failure
-            ):
-                instance.report_engine_event(
-                    "Not retrying run since it failed due to an asset or op failure and run retries "
-                    "are configured with retry_on_asset_or_op_failure set to false.",
-                    run,
-                )
-            else:
-                yield (run, retry_number)
+        if get_boolean_tag_value(run.tags.get(WILL_RETRY_TAG), default_value=False):
+            retry_number = int(run.tags.get(RETRY_NUMBER_TAG, "0")) + 1
+            yield (run, retry_number)
```

Comment:
claire's comment https://github.com/dagster-io/dagster/pull/25771#discussion_r1844654613 

> We should delete that dagster/run_will_retry tag after the retry kicks off, right? So that a get_runs call filtering on that tag will only fetch runs that are about to retry, and not runs that have already retried

Reply:
A point in favor of this approach is how the backfill daemon will use the `will_retry` tag. 

When determining if the backfill is over we need to iterate over all failed runs to see if they will be retried. If we don't remove the `will_retry` tag and we only look for the truthiness of the `will_retry` tag the backfill will never complete since a failed run that will retry will always have that tag, even if the retry has been kicked off. So to fix that we would need to fetch the run group and see if the retry has been kicked off for that run. This doesn't seem like a huge problem, but is some additional logic to keep track of.

If we remove the tag after the retry is submitted, then we don't need to add that additional info. But with the current implementation that would leave us with:
- run that was never planned to be retried has tag `dagster/will_retry=false`
- run that will be retried has tag `dagster/will_retry=true
- run that was retried has no tag
which is a bit strange

Another option is to rename the tag `dagster/waiting_to_retry` tag that moves from `True` to `False` after the retry is submitted. I think that is a bit easier to follow

Reply:
Other option could be to add the run id of the retried run. So a run would start with `dagster/will_retry=true` and we would know if the retry has been launched if a `dagster/child_run=run_id` tag was added

Reply:
I am strongly in favor of adding some sort of `child_run` tag functionality. Any process waiting on the completion of the run + it's retries will be implemented most efficiently using a tag like that.

Reply:
Perhaps you were making this point already, but just to make it explicit, you actually don't need to delete the will_retry tag if you have the child_run tag right

Reply:
to close the loop on this based on our convo in slack - we're gonna add a did_retry tag once the retry has been kicked off